cylc/uiserver/data_store_mgr.py

```diff
@@ -152,37 +191,32 @@ async def connect_workflow(self, w_id, contact_data):
         blocking the main loop.
 
         """
-        if self.loop is None:
-            self.loop = asyncio.get_running_loop()
-
         # don't sync if subscription exists
-        if w_id in self.w_subs:
+        if w_id in self.active_subscriptions:
             return
 
         self.delta_queues[w_id] = {}
 
-        # Might be options other than threads to achieve
-        # non-blocking subscriptions, but this works.
-        self.executor.submit(
-            self._start_subscription,
-            w_id,
-            contact_data['name'],
-            contact_data[CFF.HOST],
-            contact_data[CFF.PUBLISH_PORT]
-        )
-        successful_updates = await self._entire_workflow_update(ids=[w_id])
-
-        if w_id not in successful_updates:
-            # something went wrong, undo any changes to allow for subsequent
-            # connection attempts
-            self.log.info(f'failed to connect to {w_id}')
+        try:
+            # start the subscriber to keep this store updated
+            self._start_subscription(
+                w_id,
+                contact_data['name'],
+                contact_data[CFF.HOST],
+                contact_data[CFF.PUBLISH_PORT],
+            )
+            # make a one-off request to provide the initial data
+            await self._entire_workflow_update(w_id)
+        except WorkflowStopped:
+            self.disconnect_workflow(w_id)
+        except Exception as exc:
+            self.log.error(f'Failed to connect to {w_id}: {exc}')
             self.disconnect_workflow(w_id)
```

Comment:
Moved error handling up a level from the `_start_subscription` and `_entire_workflow_update` methods, no more threads to interfere with error reporting.